src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs

```diff
@@ -23,59 +23,138 @@
 namespace AspNet.Security.OAuth.Introspection {
     public class OAuthIntrospectionHandler : AuthenticationHandler<OAuthIntrospectionOptions> {
         protected override async Task<AuthenticateResult> HandleAuthenticateAsync() {
-            string header = Request.Headers[HeaderNames.Authorization];
-            if (string.IsNullOrEmpty(header)) {
-                return AuthenticateResult.Fail("Authentication failed because the bearer token " +
-                                               "was missing from the 'Authorization' header.");
-            }
+            try {
+                // Give application opportunity to find from a different location, adjust, or reject token
+                var accessTokenReceivedContext = new AccessTokenReceivedContext(Context, Options);
+
+                // event can set the token
+                await Options.Events.AccessTokenReceived(accessTokenReceivedContext);
+                if (accessTokenReceivedContext.HandledResponse) {
+                    return AuthenticateResult.Success(accessTokenReceivedContext.Ticket);
+                }
+                if (accessTokenReceivedContext.Skipped) {
```

Comment:
Hey, you told @DovydasNavickas that you'd format things in whatever code style you preferred when you merged, so I wasn't worrying about whitespace or brackets :wink: